R/tags.R

```diff
@@ -698,7 +713,35 @@ findDependencies <- function(tags, tagify = TRUE) {
     }
   }
   childDeps <- unlist(lapply(children, findDependencies, tagify = FALSE), recursive = FALSE)
-  c(childDeps, if (!is.null(dep)) dep)
+  c(childDeps, if (!is.null(deps)) deps)
+}
+
+resolveFunctionalDependencies <- function(deps) {
+  if (!length(deps)) {
+    return(deps)
+  }
+  deps <- lapply(deps, function(x) {
+    if (is_html_dependency(x)) {
+      return(list(x))
+    }
+    if (!inherits(x, "shiny.tag.function")) {
+      stop(call. = FALSE, "HTML dependencies must be a list containing either ",
+           "htmlDependency()s or tagFunction()s")
+    }
+    y <- x()
+    if (is_html_dependency(y)) {
+      return(list(deps))
+    }
+    if (!all(vapply(y, is_html_dependency, logical(1)))) {
```

Comment:
It's not entirely clear to me: if `y` is a non-list, will it enter this block and give the error message? Or will it throw some sort of (confusing) error?

Reply:
If `y` is zero-length it won't enter this block (but the `unlist()` should remove any zero-length return value, which is by design)

If `y` is non-list with positive length, it will enter this block and give the error message.

Now that I think of it, I think we need to handle the `isTag(y)` and `isTagList(y)` cases as well...